solutions/ojuz-knapsack.mdx

```diff
@@ -0,0 +1,171 @@
+---
+id: ojuz-knapsack
+source: Singapore NOI
+title: knapsack
+author: Kevin Sheng
+---
+
+## Explanation
+
+Note that the bound on $S$ is rather small and that $1 \leq W_i \leq S$, so we can group the items together by weight.  
+After this, we can implement the knapsack problem on the grouped together items. We also sort items of the same weight by reverse value, as it is always optimal to select the item with greater value among two items of the same weight.  
+
+## Implemention
+
+**Time Complexity:** $\mathcal{O}(S^2 \cdot \log S)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <vector>
+#include <map>
+#include <algorithm>
+
+using std::cout;
+using std::endl;
+using std::vector;
+using std::pair;
+
+int main() {
+	int limit;
+	int type_num;
+	std::cin >> limit >> type_num;
+	
+	std::map<int, vector<pair<int, int>>> by_weight;
+	for (int t = 0; t < type_num; t++) {
+		int value;
+		int weight;
+		int amt;
+		std::cin >> value >> weight >> amt;
+		if (weight <= limit && amt > 0) {
+			by_weight[weight].push_back({value, amt});
+		}
+	}
+	
+	/*
+	 * best[i][j] contains the most value we can
+	 * get using j weight and the first i weight types
+	 */
+	vector<vector<long long>> best(
+		by_weight.size() + 1, vector<long long>(limit + 1, INT32_MIN)
+	);
```

Comment:
Think the standard name for dp arrays is just ```f``` or ```dp```.

Reply:
it's more self-explanatory tbh